contracts/TaskBook.sol

```diff
@@ -5,83 +5,154 @@ import './AccountManager.sol';
+	event SolutionRevealed(uint taskID, uint randomBits);
+	event TaskStateChange(uint taskID, uint state);
 
 	struct Task {
 		address owner;
-		address[] solvers;
 		address selectedSolver;
 		uint minDeposit;
 		bytes32 taskData;
-		uint numSolvers;
-		address[] challengers;
-		uint numChallengers;
+		mapping(address => bytes32) challenges;
+		uint state;
+		bytes32 blockhash;
+		bytes32 randomBitsHash;
+	}
+
+	struct Solution {
+		bytes32 solutionHash0;
+		bytes32 solutionHash1;
+		bool solution0Correct;
+		address[] solution0Challengers;
+		address[] solution1Challengers;
 	}
 
+	mapping(uint => Task) private tasks;
+	mapping(uint => Solution) private solutions;
+
 	//Task Issuers create tasks to be solved
 	function createTask(uint minDeposit, bytes32 taskData, uint numBlocks) returns (bool) {
 		require(balances[msg.sender] >= minDeposit);
-		tasks[numTasks] = Task(msg.sender, new address[](maxSolvers), 0x0, minDeposit, taskData, 0, new address[](maxChallengers), 0);
+		Task storage t = tasks[numTasks];
+		t.owner = msg.sender;
+		t.minDeposit = minDeposit;
+		t.taskData = taskData;
+		tasks[numTasks] = t;
 		log0(sha3(msg.sender));//possible bug if log is after event
 		TaskCreated(numTasks, minDeposit, block.number+numBlocks);
 		numTasks++;
 		return true;
 	}
 
-	//Solver registers for tasks
-	function registerForTask(uint taskID, uint minDeposit, bytes32 randomBitsHash) returns(bool) {
-		require(balances[msg.sender] >= minDeposit);
-		Task t = tasks[taskID];
-		require(!(t.owner == 0x0));
-		require(t.numSolvers < maxSolvers);
-		solverRandomBitsHash[msg.sender][taskID] = randomBitsHash;
-		//random = sha3(random, block.blockhash(block.number-1));
-		t.solvers[t.numSolvers] = msg.sender;
-		t.numSolvers++;
-		log0(bytes32(sha3(msg.sender)));
+	function changeTaskState(uint taskID, uint newState) returns (bool) {
+		Task storage t = tasks[taskID];
+		require(t.owner == msg.sender);
+		t.state = newState;
+		TaskStateChange(taskID, newState);
 		return true;
 	}
 
-	//Task Issuer tells TrueBitExchange to select a solver
-	function selectSolver(uint taskID) returns (bool) {
-		Task t = tasks[taskID];
-		require(msg.sender == t.owner);
-		//uint random = uint(block.blockhash(block.number-1));
-		//address randomSolver = t.solvers[uint(random) % t.numSolvers];
-		address solver = t.solvers[0];
-		t.selectedSolver = solver;
-		SolverSelected(taskID, solver, t.taskData, t.minDeposit);
+	//Solver registers for tasks, if first to register than automatically selected solver
+	//0->1
+	function registerForTask(uint taskID, bytes32 randomBitsHash) returns(bool) {
+		Task storage t = tasks[taskID];
+		require(balances[msg.sender] >= t.minDeposit);
+		require(!(t.owner == 0x0));
+		require(t.state == 0);
+		require(t.selectedSolver == 0x0);
+		t.selectedSolver = msg.sender;
+		t.randomBitsHash = randomBitsHash;
+		t.blockhash = block.blockhash(block.number-1);
```

Comment:
The reason I think it should only store the block number, and then get the block hash later is that the solver knows the blockhash for previous blocks and can therefore easily generate bits to get the forced error.
The reason I think it should be this block hash, and not a block hash selected by task giver is that if task giver and solver are the same person, and timeout would be 10 blocks, then in this case there would be 10 times better probability to get a block with forced error.

Reply:
Doesn't the decision on whether a forced-error is in effect depend on the `block hash` _after_ the solver submits their two solutions?

Reply:
Currently it is calculated as at line 133:
`if(uint(sha3(randomBits, tasks[taskID].blockhash)) < forcedErrorThreshold)`

Yes, it will have to depend on some block hash that comes after the random bits have been committed. My first suggestion is that it depends on the hash of the block that includes the commit transaction.